### PR TITLE
posix: Refactor pthread_join() to remove dead code warnings

### PR DESCRIPTION
**Pull request by finikorg**
Fixes "dead code" warnings, the logging was after `continue` operation in K_SPINLOCK()

### CODE REVIEW

File: lib/posix/pthread.c [hunk 1]
```diff
@@ -667,15 +667,15 @@ int pthread_join(pthread_t pthread, void **status)
 	{
 		if (t->detachstate != PTHREAD_CREATE_JOINABLE) {
 			ret = EINVAL;
-			K_SPINLOCK_BREAK;
 			LOG_ERR("Pthread %p is not a joinable", &t->thread);
+			K_SPINLOCK_BREAK;
```
Comment (cfriedt) [inline lib/posix/pthread.c:671]: Ack yeah, good catch. It's probably not great to log with the spinlock held though.

Is it possible to defer the call to log_err ? Maybe just by using a `const char *` that is initialized to `NULL`?

**Review by cfriedt — CHANGES_REQUESTED** [review]

File: lib/posix/pthread.c (hunk 1 above)
Comment (finikorg) [reply lib/posix/pthread.c:671]: Maybe remove these logs at all?

Comment (cfriedt) [reply lib/posix/pthread.c:671]: Sure

**Review by cfriedt — DISMISSED** [review]

File: tests/subsys/mgmt/mcumgr/mcumgr_client/src/img_gr_stub.c [hunk 2]
```diff
@@ -242,7 +242,7 @@ void img_upload_init_verify(struct net_buf *nb)
 	image = UINT32_MAX;
 
 	rc = zcbor_map_decode_bulk(zsd, list_res_decode, ARRAY_SIZE(list_res_decode), &decoded);
-	if (rc || data.len == 0 || offset == SIZE_MAX || image != TEST_IMAGE_NUM) {
+	if (rc) {
```
Comment (de-nordic) [inline tests/subsys/mgmt/mcumgr/mcumgr_client/src/img_gr_stub.c:245]: Why? The `zcbor_map_decode_bulk` will return ok even if the decoded fields are not found.
The check here tests if frame, that has been correctly decoded, contains all needed data.

**Review by de-nordic — DISMISSED** [review]

**Comment by nashif** [comment]
the title of the PR is misleading.

File: lib/posix/pthread.c (hunk 1 above)
Comment (finikorg) [reply lib/posix/pthread.c:670]: > Ack yeah, good catch. It's probably not great to log with the spinlock held though.
> 
> Is it possible to defer the call to log_err ? Maybe just by using a `const char *` that is initialized to `NULL`?

like this? #0901f63

**Review by cfriedt — APPROVED** [review]

File: lib/posix/pthread.c [hunk 3]
```diff
@@ -685,13 +682,20 @@ int pthread_join(pthread_t pthread, void **status)
...
-	if (ret != 0) {
+	switch (ret) {
+	case ESRCH:
+		LOG_ERR("Pthread %p has already been joined", &t->thread);
 		return ret;
+	case EINVAL:
+		LOG_ERR("Pthread %p is not a joinable", &t->thread);
+		return ret;
+	case 0:
+		break;
```
Comment (de-nordic) [inline lib/posix/pthread.c:693]: Just wondering whether it wouldn't be as good to just do:
```
LOG_ERR("Pthraead %p %s", &t->thread, msg)
```
and set msg according to ret. Probably would save us some flash.

**Review by de-nordic — APPROVED** [review]

File: lib/posix/pthread.c (hunk 3 above)
Comment (cfriedt) [reply lib/posix/pthread.c:693]: yep, that would work. would probably be fewer lines of code too

**Pull request merged**